contracts/MultiPartyEscrow.sol

```diff
@@ -93,23 +95,53 @@ contract MultiPartyEscrow {
         require(balances[msg.sender] >= value, "Insufficient balance in the contract.");
         require(signer != address(0));
 
+        require(_openChannel(msg.sender, signer, recipient, groupId, value, expiration), "Unable to open channel");
+        return true;
+    }
+    
+    //open a channel on behalf of the user. Sender should send the signed permission to open the channel
+    function openChannelByThirdParty(address sender, address signer, address recipient, bytes32 groupId, uint256 value, uint256 expiration, uint256 messageNonce, uint8 v, bytes32 r, bytes32 s) 
+    public
+    returns(bool) 
+    {
+        require(balances[sender] >= value, "Insufficient balance");
+
+        require(messageNonce >= block.number-5 && messageNonce <= block.number+5, "Invalid message nonce");
```

Comment:
Maing messageNonce equal to block.number is a nice idea, but I'm affraid it will not work in real Ethereum network. Mining can take much longer then 5 blocks... 

Actually we probably do not need nonce at all, because user will never need to open two channel for the same recipient+groupId.

If we still want to keep the theoretical possibility to open several channels for the same recipient+groupId then simple random nonce is enough.

Reply:
Yes, there is a probability it might take longer time than 5 blocks esp during network congestion. We have come up with this idea to see how best we can eliminate the storage of messages in the contract by checking with +/- 5 blocks.
I am thinking of passing messageNonce as a blocknumber and it will be very rare chances that the two tx will be in the same block from the same user account.
We need to have the possibility to have multiple channels for the same recipient + groupId. May be in certain scenarios we might needed.

Reply:
But in the current version you do store message, and it is not so expensive after all. In any case you will need to store something in order to prevent replay attack. So I don't see how you will make it cheaper. As I know, it is cheaper to keep information in the mapping then to remove it, so I think the current solution with usedMessages is ok. 

In which scenario we need multiply channels for the same sender+signer+recipient+groupId? In any case, if we need this possiblity we should keep messageNonce, but we shouldn't verify it (user should user random messageNonce)